Alerting/Docs: Non-timeseries / numeric alerting (#35190)
grafana/grafana: docs/sources/alerting/unified-alerting/grafana-managed-number-alerts.md

+++

> title = "Numeric data Grafana manged alert rules"

Reviewer: ```suggestion
title = "Numeric data Grafana managed alert rules"
```

> description = "Number data based Grafana manged alert rules"

Reviewer: ```suggestion
description = "Number data based Grafana managed alert rules"
```

keywords = ["grafana", "alerting", "guide", "rules", "create"]
weight = 400
+++

# Alerting on numeric data

> With certain data sources number data that is not time series can be directly alerted on, or passed into Server Side Expressions (SSE). This allows for more efficiency by doing more processing within the data source, and can also simplify alert rules.
> 

Reviewer 2: ```suggestion
Among certain data sources numeric data that is not time series can be directly alerted on, or passed into Server Side Expressions (SSE). This allows for more processing and resulting efficiency within the data source, and it can also simplify alert rules.
```

When alerting on numeric data instead of time series data, there is no need to reduce each labeled time series into a single number. Instead labeled numbers are returned to Grafana instead.

## Tabular Data

This feature is supported with backend data sources that query tabular data:

- SQL data sources such as MySQL, Postgres, MSSQL, and Oracle.
- The Azure Kusto based services: Azure Monitor (Logs), Azure Monitor (Azure Resource Graph), and Azure Data Explorer.

A query with Grafana managed alerts or SSE is considered numeric with these data sources, if:

- The "Format AS" option is set to "Table" in the data source query.
- The table response returned to Grafana from the query includes only one numeric (e.g. int, double, float) column, and optionally additional string columns.

If there are string columns then those columns become labels. The name of column becomes the label name, and the value for each row becomes the value of the corresponding label. If multiple rows are returned, then each row should be uniquely identified their labels.

## Example

> If you have a MySQL table called "DiskSpace" like the following:

Reviewer: It would be nice to include also a description of this alert like: fire when at least one server disk has lower disk space than 5%.

> If you have a MySQL table called "DiskSpace" like the following:

Reviewer 2: ```suggestion
For a MySQL table called "DiskSpace":
```

| Time        | Host | Disk | PercentFree
| ----------- | ---  | -----| --------
| 2021-June-7 | web1 | /etc | 3
| 2021-June-7 | web2 | /var | 4
| 2021-June-7 | web3 | /var | 8
| ...         | ...  | ...  | ...

> You can query this data filtering on time, but without returning time series to Grafana:

Reviewer 2: ```suggestion
You can query the data filtering on time, but without returning time series to Grafana. For example:
```

```sql
SELECT Host, Disk, CASE WHEN PercentFree < 5.0 THEN PercentFree ELSE 0 END FROM (
  SELECT 
      Host, 
      Disk, 
      Avg(PercentFree) 
  FROM DiskSpace
  Group By 
    Host, 
    Disk
  Where __timeFilter(Time)
```

> This query would return the following Table response to Grafana:

Reviewer 2: ```suggestion
This query returns the following Table response to Grafana:
```

| Host | Disk | PercentFree
| ---  | -----| --------
| web1 | /etc | 3
| web2 | /var | 4
| web3 | /var | 0

> When this query is used as the **condition** in an alert rule, the non-zero will be alerting. So three alert instances will be produced:

Reviewer 2: ```suggestion
When this query is used as the **condition** in an alert rule, then the non-zero will be alerting. As a result, three alert instances are produced:
```

+++

> title = "Numeric data Grafana managed alert rules"

Reviewer 2: How about "Alert rules for numeric data"? The title as-is is a little confusing. If we want to stress that is a Grafana alert, then: "Grafana managed alert rules for numeric data".
